Handle Missing Required Fields in a Non-Fatal Way

mcdonnnj opened the pull request:
# <!--- Provide a general summary of your changes in the Title above -->

## 🗣 Description

This PR adds functionality to check if required fields are missing from an assessment record. It will then log a warning and pass to the next record if any are missing.
<!--- Describe your changes in detail -->

## 💭 Motivation and Context

Currently if a required field is missing the lambda will die in a fatal manner. This is undesirable behavior.
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## 🧪 Testing

Automated tests pass. I tested the added snippet on its own with test data.
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## 📷 Screenshots (if appropriate)

## 🚥 Types of Changes

<!--- What types of changes does your code introduce? -->
<!--- Put an `x` in all the boxes that apply: -->

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (causes existing functionality to change)

## ✅ Checklist

<!--- Go over all the following points, and put an `x` in all the
boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask.
We're here to help! -->

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

mcdonnnj committed:
Explicitly check for required fields and handle them

The lambda in production fails when a required field is missing. This updates
the lambda to warn about missing requied fields and move to the next record
instead.

mcdonnnj committed:
Change pass -> continue to mirror tested code

dav3r commented on adi/assessment_data_import.py:
```diff
@@ -147,6 +147,32 @@ def import_data(
         for assessment in assessment_data:
+            # Check for the most required of fields
+            if "id" not in assessment:
+                logging.warning("Assessment missing 'id'! Skipping...")
```
When the `id` is missing, we should output the assessment info that is present so that a human can figure out what went wrong.

mcdonnnj replied:
I only omitted record information because I wasn't sure if we wanted assessment information showing up in the CloudWatch log stream. If we didn't want that information in the log stream I could include the record indexes and output that information. That way a human could retrieve the JSON from S3 and look at the specified record index. Otherwise I can just output the record if it's missing an `id` key.

dav3r replied:
Agreed- the record index should be sufficient.  Thanks!

mcdonnnj replied:
Please see e3a4b3e for adding the record index to the warning message.

dav3r reviewed (changes requested):
This is an excellent improvement, but I have one small logging change to request.

jsf9k commented on adi/assessment_data_import.py:
```diff
@@ -147,6 +147,32 @@ def import_data(
+            ]
+            if missing_fields:
+                logging.warning(
+                    f"'{assessment['id']}' is missing the following required field(s):"
```
I thought we had decided to use the printf-style logging jazz everywhere, since that provides a little bit of optimization.

mcdonnnj replied:
I was keeping it consistent with the rest of the lambda's code. I'm happy to change all the logging to reflect that though.

mcdonnnj replied:
Please see 6c83cda for changing all logging to use the printf-style.

jsf9k reviewed (commented):
I have a comment for your consideration.  Other than that, I am happy with these changes once you and @dav3r resolve the issue he brought up.  Just re-request my review when that is settled.

mcdonnnj committed:
Change logging messages to use printf style

This change aligns with general best practices for Python logging. The logging
system will only interpolate strings from a log level currently being logged.

mcdonnnj committed:
Add record index to warning about assessment missing 'id' field

dav3r reviewed (approved):
Looks great - nice work! 🦅

jsf9k reviewed (approved):
Strong work was done here!
